Update member_joined_channel events to be idempotent

joekr opened the pull request:
Now using a database to ensure that multiple events sent to the API won't trigger multiple welcome messages.

joekr committed:
Update member_joined_channel events to be idempotent

Now using a database to ensure that multiple events sent to the API
won't trigger multiple welcome messages.

joekr committed:
Fix github action issue

Build failed with `F821 undefined name 'fdk'`

xtoddx commented on events/func.py:
```diff
@@ -1,35 +1,115 @@
+        if len(query_resp.data.items) == 0:
+            should_post = True
+    except Exception as ex:
+        should_post = False
```
This would already be false from line 32.

Don't forget that in python empty lists don't pass `if`, so `if len(x) == 0` is the same as `if not x`. But hey, if you charge for character then I say go for it.

xtoddx commented on events/func.py:
```diff
@@ -1,35 +1,115 @@
+    if local_dev:
+        try:
+            await slack_client.chat_postMessage(
+                channel=f"#{welcome_channel}",
```
Why not reuse the channel from the event, or random_channel_id?

joekr replied:
Well mainly so I could test, but you are right. I will set it to use the channel ID since I already check to make sure it is the random channel.

xtoddx commented:
I think you're still going to have raciness because the multiple events will be handled by different invocations of the function, and all can read an empty set before the first write completes.

You could make each try writes that conflict and only have the one post to slack after writing the DB w/o conflicts.

joekr commented:
Great point. I will need to retool the DB a bit and add a unique index.

joekr committed:
Update how checking db for event is handled

To make sure we don't have a race condition the "check" will be to
write to the DB. If the DB returns an `existing_version` then another
function has handled the event.

I've also added a quick `test_event_idempotent.sh` script to help
test locally.

joekr commented:
So it turns out the [NosqlClient](https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/nosql/client/oci.nosql.NosqlClient.html#oci.nosql.NosqlClient) client has an [update_row](https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/nosql/client/oci.nosql.NosqlClient.html#oci.nosql.NosqlClient.update_row) method that takes a [UpdateRowDetails](https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/nosql/models/oci.nosql.models.UpdateRowDetails.html#oci.nosql.models.UpdateRowDetails) model.

In this model you can set `option="IF_ABSENT"`. If you set that then after calling `update_row` the returned [UpdateRowResult](https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/nosql/models/oci.nosql.models.UpdateRowResult.html#oci.nosql.models.UpdateRowResult) will contain an attribute [existing_version](https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/nosql/models/oci.nosql.models.UpdateRowResult.html#oci.nosql.models.UpdateRowResult.existing_version) other than `None` to denote that the put "failed". 

I'm using this to check whether or not another function invocation already handled the event.

joekr committed:
Fix issue with local dev

This would have prevented the function from sending an actual message.

arthall commented on events/func.py:
```diff
@@ -2,34 +2,105 @@
+    # If write fails event already handled
+    should_post = update_db(oci_client, event)
+
+    return should_post
```
nit: Are there plans to add to should_post_message? If not, this function can be removed and update_db can be called directly.

joekr replied:
Fixed

arthall commented on events/func.py:
```diff
@@ -2,34 +2,105 @@
+    if (event["channel"] != random_channel_id):
+        return
+
+    oci_client = None
+    if local_dev:
+        config = oci.config.from_file("./oci_config")
+        oci_client = oci.nosql.NosqlClient(config)
+    else:
+        signer = oci.auth.signers.get_resource_principals_signer()
+        oci_client = oci.nosql.NosqlClient({}, signer=signer)
```
nit: This could be made into a function and the results saved to oci_client. Something along the lines of 
`oci_client = init_client(local_dev)`.  additional nit: the client is a nosql client so `nosql_client` or `db_client` might help someone quickly understand what the client is doing.

joekr replied:
Fixed

joekr committed:
Cleanup some functions based on feedback
